README
Homebridge plugin for PushCut notifications.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 1.1.4

### Fixes

- Removed external dependencies for URL checks. These caused install problem or long install times.
- Using built-in URL instead.

### Bump dependencies

- Node v18.15.0 LTS
- Homebridge v1.6.0
- homebridge-lib v6.3.12
- luxon v3.3.0
- events v3.3.0

## 1.1.3

### Enhancements

- Allow the values to be restored from the configuration file after restart.
- Selection of sounds implemented.
- Selection of devices that will receive notifications.
- Log level is working.

### Bump dependencies

- Node v16.17.0 LTS
- homebridge-lib v5.6.7
- luxon v3.0.3

## 1.1.2

### Bump dependencies

- homebridge v1.5.0
- homebridge-lib v5.6.1
- Node v16.15.1 LTS
- luxon v2.4.0
- parse-url v5.0.8
- is_url_http v2.3.2
- events v3.3.0
- np v7.6.2

## 1.1.1

### Bump dependencies

- homebridge v1.4.0
- homebridge-lib v5.2.0
- Node v16.13.2 LTS
- luxon v2.2.0
- parse-url v5.0.7
- is_url_http v2.2.7
- np v7.6.0
- Moved np to devDependencies


lib/PcTypes.js
// homebridge-pushcutter/lib/PcTypes.js
// Copyright © 2021 Kenneth Jagenheim. All rights reserved.
//
// Custom HomeKit Characteristics and common functions.

const homebridgeLib = require('homebridge-lib')

const sleep = (seconds) => {
  return new Promise((resolve, reject) => {
    setTimeout(resolve, seconds * 1000)
  })
}

function wait(ms, opts = {}) {
  return new Promise((resolve, reject) => {
    let timerId = setTimeout(resolve, ms);
    if (opts.signal) {
      // implement aborting logic for our async operation
      opts.signal.addEventListener('abort', event => {
        clearTimeout(timerId);
        reject(event);
      })
    }
  })
}

function stringInject(str, data) {
  if (typeof str === 'string' && (data instanceof Array)) {
    return str.replace(/({\d})/g, function (i) {
      return data[i.replace(/{/, '').replace(/}/, '')];
    });
  } else if (typeof str === 'string' && (data instanceof Object)) {
    if (Object.keys(data).length === 0) {
      return str;
    }
    for (let key in data) {
      return str.replace(/({([^}]+)})/g, function (i) {
        let key = i.replace(/{/, '').replace(/}/, '');
        if (!data[key]) {
          return i;
        }
        return data[key];
      });
    }
  } else if (typeof str === 'string' && data instanceof Array === false || typeof str === 'string' && data instanceof Object === false) {
    return str;
  } else {
    return false;
  }
}

const regExps = {
  uuid: /^[0-9A-F]{8}-[0-9A-F]{4}-[1-5][0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/,
  uuidPrefix: /^[0-9A-F]{1,8}$/,
  uuidSuffix: /^-[0-9A-F]{4}-[1-5][0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/
}

function uuid(id, suffix = '-0000-1000-8000-656261617577') {
  //'-0000-1000-8000-0026BB765291') {
  if (typeof id !== 'string') {
    throw new TypeError('id: not a string')
  }
  if (!regExps.uuidPrefix.test(id)) {
    throw new RangeError(`id: ${id}: invalid id`)
  }
  if (typeof suffix !== 'string') {
    throw new TypeError('suffix: not a string')
  }
  if (!regExps.uuidSuffix.test(suffix)) {
    throw new RangeError(`suffix: ${suffix}: invalid suffix`)
  }
  return ('00000000' + id).slice(-8) + suffix
}

class PcTypes extends homebridgeLib.CustomHomeKitTypes {
  constructor(homebridge) {
    super(homebridge)

    this.createCharacteristicClass('Mute', uuid('0A5'), {
      format: this.Formats.UINT16,
      unit: this.Units.SECONDS,
      minValue: 1,
      maxValue: 3600,
      perms: [this.Perms.READ, this.Perms.NOTIFY, this.Perms.WRITE],
    }, 'Mute time')

    this.createCharacteristicClass('PcTitleDisplay', uuid('0A6'), {
      format: this.Formats.STRING,
      perms: [this.Perms.READ, this.Perms.NOTIFY],
    }, 'Pushcut title (current)')

    this.createCharacteristicClass('PcTextDisplay', uuid('0A7'), {
      format: this.Formats.STRING,
      perms: [this.Perms.READ, this.Perms.NOTIFY],
    }, 'Pushcut text (current)')

    this.createCharacteristicClass('PcTitle', uuid('0B6'), {
      format: this.Formats.STRING,
      perms: [this.Perms.READ, this.Perms.NOTIFY, this.Perms.WRITE],
    }, 'Pushcut title (enter new)')

    this.createCharacteristicClass('PcText', uuid('0B7'), {
      format: this.Formats.STRING,
      perms: [this.Perms.READ, this.Perms.NOTIFY, this.Perms.WRITE],
    }, 'Pushcut text (enter new)')

    this.createCharacteristicClass('Random', uuid('0A9'), {
      format: this.Formats.BOOL,
      perms: [this.Perms.READ, this.Perms.NOTIFY, this.Perms.WRITE],
    }, 'Random enabled')

    this.createCharacteristicClass('SoundDisplay', uuid('0AA'), {
      format: this.Formats.STRING,
      perms: [this.Perms.READ, this.Perms.NOTIFY],
    }, 'Notification sound (current)')

    this.createCharacteristicClass('Sound', uuid('0AB'), {
      format: this.Formats.STRING,
      perms: [this.Perms.READ, this.Perms.NOTIFY, this.Perms.WRITE],
    }, 'Notification sound (set new)')

    this.createCharacteristicClass('DevicesDisplay', uuid('0AC'), {
      format: this.Formats.STRING,
      perms: [this.Perms.READ, this.Perms.NOTIFY],
    }, 'Notificattion devices (current)')

    this.createCharacteristicClass('Devices', uuid('0AD'), {
      format: this.Formats.STRING,
      perms: [this.Perms.READ, this.Perms.NOTIFY, this.Perms.WRITE],
    }, 'Notification devices (set new)')

    this.createCharacteristicClass('Default', uuid('0B8'), {
      format: this.Formats.BOOL,
      perms: [this.Perms.READ, this.Perms.NOTIFY, this.Perms.WRITE],
    }, 'Restore default')

  }
}

const validSounds = [
  "default",
  "vibrateOnly",
  "system",
  "subtle",
  "question",
  "jobDone",
  "problem",
  "loud",
  "lasers"
]

module.exports.PcTypes = PcTypes
module.exports.sleep = sleep
module.exports.wait = wait
module.exports.stringInject = stringInject
module.exports.validSounds = validSounds

lib/PcAccessory.js
// homebridge-pushcutter/lib/PcAccessory.js
// Copyright © 2021 Kenneth Jagenheim. All rights reserved.
//
// Homebridge plugin for Pushcut notifications.

'use strict';

const homebridgeLib = require('homebridge-lib');
const SwitchService = require('./SwitchService');

class PcAccessory extends homebridgeLib.AccessoryDelegate {
  constructor(platform, params) {
    super(platform, params);
    this.api = params.api;
    //    this.notificationName = params.notificationName
    this.mute = params.mute || 5;
    this.notification = params.notification;
    this.pcClient = params.pcClient;
    this.useWebhook = params.useWebhook;
    this.webhook = params.webhook;
    this.pcTitle = params.pcTitle;
    this.pcText = params.pcText;
    this.pcImage = params.pcImage;
    this.startOnReboot = params.startOnReboot;
    this.rebootTitle = params.rebootTitle;
    this.rebootText = params.rebootText;
    this.sound = params.sound;
    this.devices = params.devices;
    this.useConfig = params.useConfig;
    this.pushcutDevices = params.pushcutDevices;
    this.pc = platform.pc;

    if (this.useWebhook) {
      let host;
      let path;
      try {
        let fullUrl = new URL(this.webhook);
        host = fullUrl.host;
        path = fullUrl.pathname;
        this.log(
          'Webhook split into host %s and path %s',
          host,
          path
        );
      } catch (error) {
        this.console.error();
        'Error in webhook URL %s: %s', this.webhook, error.message;
      }
      this.pcClient = new homebridgeLib.HttpClient({
        https: true,
        host: host,
        //        headers: headers,
        json: true,
        maxSockets: 1,
        path: path,
        timeout: 5, //this.config.timeout,
        validStatusCodes: [200, 401, 403, 404],
      });
      this.notification = '/';

      this.pcClient
        .on('error', (error) => {
          this.log(
            'Webhook request %d: Error = %s %s',
            error.request.id,
            error.request.method,
            error.request.resource
          );
          this.warn(
            'Webhook request %d: Error = %d, %s',
            error.request.id,
            error.statusCode,
            error.statusMessage
          );
        })
        .on('request', (request) => {
          this.debug(
            'Webhook request %d: Request = %s, Resource = %s',
            request.id,
            request.method,
            request.resource
          );
          this.vdebug(
            'Webhook request %d: Request = %s, URL = %s',
            request.id,
            request.method,
            request.url
          );
        })
        .on('response', (response) => {
          this.vdebug(
            'Webhook request %d: Response = %j',
            response.request.id,
            response.body
          );
          this.debug(
            'Webhook request %d: Response = %d %s',
            response.request.id,
            response.statusCode,
            response.statusMessage
          );
        });
    }
    this.switchService = new SwitchService(this, {
      primaryService: true,
    });
    this.manageLogLevel(
      this.switchService.characteristicDelegate('logLevel')
    );

    this.debug('Accessory initialised');
    this.heartbeatEnabled = true;
    setImmediate(() => {
      this.emit('initialised');
    });
    this.on('heartbeat', async (beat) => {
      await this.heartbeat(beat);
    });
    this.on('shutdown', async () => {
      return this.shutdown();
    });
  }

  async shutdown() {
    this.debug('Nothing to do at shutdown');
  }

  async heartbeat(beat) {
    if (beat % this.switchService.values.heartrate === 0) {
    }
  }
}

module.exports = PcAccessory;


lib/SwitchService.js
// homebridge-pushcutter/lib/SwitchService.js
// Copyright © 2021 Kenneth Jagenheim. All rights reserved.
//
// Homebridge plugin for Pushcut notifications.

'use strict';

const homebridgeLib = require('homebridge-lib');
const AbortController = require('abort-controller');
const PcTypes = require('./PcTypes');
const wait = PcTypes.wait;
const validSounds = PcTypes.validSounds;
const replace = require('key-value-replace');
const delimiter = ['{', '}'];
const { DateTime } = require('luxon');
const pushcutDevices = [];

class SwitchService extends homebridgeLib.ServiceDelegate {
  constructor(switchAccessory, params = {}) {
    params.name = switchAccessory.name;
    params.Service = switchAccessory.Services.hap.Switch;
    super(switchAccessory, params);
    this.useApi = switchAccessory.useApi;
    this.locale = switchAccessory.api.locale || false;
    this.notification = switchAccessory.notification;
    this.pcClient = switchAccessory.pcClient;
    this.pcTitle = switchAccessory.pcTitle;
    this.pcText = switchAccessory.pcText;
    this.pcImage = switchAccessory.pcImage;
    this.mute = switchAccessory.mute;
    this.startOnReboot = switchAccessory.startOnReboot;
    this.rebootTitle = switchAccessory.rebootTitle;
    this.rebootText = switchAccessory.rebootText;
    this.sound = switchAccessory.sound;
    this.devices = switchAccessory.devices;
    this.useConfig = switchAccessory.useConfig;
    this.pushcutDevices = switchAccessory.pushcutDevices;
    this.pc = switchAccessory.pc;
    this.switchName = params.name;
    this.switchOn = false;
    this.timeUp = false;
    this.useReboot = false;

    this.addCharacteristicDelegate({
      key: 'on',
      Characteristic: this.Characteristics.hap.On,
      value: false,
    }).on('didSet', (value) => {
      this.switchOn = value;
      this.debug("Calling 'setOn' with value %s", this.switchOn);
      this.setOn();
    });

    this.addCharacteristicDelegate({
      key: 'mute',
      value: this.mute,
      Characteristic: this.pc.Characteristics.Mute,
    });

    this.addCharacteristicDelegate({
      key: 'pcTitle',
      value: this.pcTitle,
      Characteristic: this.pc.Characteristics.PcTitle,
    }).on('didSet', (value) => {
      if (value) {
        this.values.pcTitleDisplay = value;
      } else {
        this.values.pcTitleDisplay = 'Uses Pushcut defined title';
      }
      this.values.pcTitle = value;
    });

    this.addCharacteristicDelegate({
      key: 'pcTitleDisplay',
      value: this.pcTitle,
      Characteristic: this.pc.Characteristics.PcTitleDisplay,
    });

    this.addCharacteristicDelegate({
      key: 'pcText',
      value: this.pcText,
      Characteristic: this.pc.Characteristics.PcText,
    }).on('didSet', (value) => {
      if (value) {
        this.values.pcTextDisplay = value;
      } else {
        this.values.pcTextDisplay = 'Uses Pushcut defined text';
      }
      this.values.pcText = value;
    });

    this.addCharacteristicDelegate({
      key: 'pcTextDisplay',
      value: this.pcText,
      Characteristic: this.pc.Characteristics.PcTextDisplay,
    });

    this.addCharacteristicDelegate({
      key: 'sound',
      value: this.sound,
      Characteristic: this.pc.Characteristics.Sound,
    }).on('didSet', (value) => {
      if (validSounds.includes(value)) {
        this.values.sound = value;
        if (value !== 'default') {
          this.values.soundDisplay = value;
        } else {
          this.values.soundDisplay = 'Uses Pushcut defined sound';
        }
      } else {
        this.warn('Illegal sound value entered.');
      }
    });

    this.addCharacteristicDelegate({
      key: 'soundDisplay',
      value: this.sound,
      Characteristic: this.pc.Characteristics.SoundDisplay,
    });

    this.addCharacteristicDelegate({
      key: 'devices',
      value: this.devices,
      Characteristic: this.pc.Characteristics.Devices,
    }).on('didSet', (value) => {
      if (value) {
        let deviceArray = value.split(',');
        let deviceValid = true;
        deviceArray.forEach((device) => {
          if (!this.pushcutDevices.includes(device)) {
            deviceValid = false;
          }
        });
        if (deviceValid) {
          this.values.devices = value;
          this.values.devicesDisplay = value;
        } else {
          this.warn('Illegal device name(s) entered.');
        }
      } else {
        this.values.devicesDisplay = 'Send to all Pushcut devices';
      }
    });

    this.addCharacteristicDelegate({
      key: 'devicesDisplay',
      value: this.devices,
      Characteristic: this.pc.Characteristics.DevicesDisplay,
    });

    this.addCharacteristicDelegate({
      key: 'heartrate',
      Characteristic: this.Characteristics.my.Heartrate,
      props: {
        minValue: 1,
        maxValue: 60,
        minStep: 1,
      },
      value: 15,
    });

    this.addCharacteristicDelegate({
      key: 'logLevel',
      Characteristic: this.Characteristics.my.LogLevel,
      value: switchAccessory.platform.logLevel,
      //    }).on('didSet', (value) => {
      //      switchAccessory.logLevel = value
    });

    this.addCharacteristicDelegate({
      key: 'setDefault',
      value: false,
      Characteristic: this.pc.Characteristics.Default,
    }).on('didSet', (value) => {
      this.setConfigValues();
    });

    if (this.useConfig) {
      this.log('Restoring values from config file');
      this.setConfigValues();
    }

    this.values.on = false;
    if (this.startOnReboot) {
      this.initSwitch();
    }
  }

  async setConfigValues() {
    this.values.pcTitle = this.pcTitle;
    this.values.pcText = this.pcText;
    this.values.pcImage = this.pcImage;
    this.values.sound = this.sound;
    this.values.devices = this.devices;
    this.values.startOnReboot = this.startOnReboot;
    await wait(500);
    this.values.setDefault = false;
  }

  async initSwitch() {
    await wait(5000);
    this.log('Send notification at reboot');
    this.useReboot = true;
    this.values.on = true;
  }

  async setOn() {
    let delayType = '';
    let json = {};

    if (!this.switchOn) {
      if (!this.timeUp) {
        try {
          this.debug('Aborting the mute period before time is up');
          this.ac.abort();
        } catch (e) {
          this.warn('Error when aborting mute: %s', e.message);
        }
      }
    } else {
      // Wait to ensure that updated values are set before the notification
      await wait(500);
      this.ac = new AbortController();
      this.signal = this.ac.signal;

      let json = {};
      let jsonObject = {};
      let dt = DateTime.now();
      if (this.locale) {
        dt = dt.setLocale(this.locale);
      }
      let textObject = {};
      textObject.time = dt.toLocaleString(DateTime.TIME_SIMPLE);
      textObject.date = dt.toLocaleString(
        DateTime.DATE_MED_WITH_WEEKDAY
      );
      textObject.week = dt.toFormat('WW');
      textObject.name = this.switchName;

      let pcTitle = this.useReboot
        ? this.rebootTitle
        : this.values.pcTitle;
      let pcText = this.useReboot
        ? this.rebootText
        : this.values.pcText;
      if (pcTitle) {
        pcTitle = replace(pcTitle, textObject, delimiter);
        this.debug('%o, %s', textObject, pcTitle);
        jsonObject = { title: pcTitle };
        json = Object.assign(json, jsonObject);
      }
      if (pcText) {
        pcText = replace(pcText, textObject, delimiter);
        this.debug('%o, %s', textObject, pcText);
        jsonObject = { text: pcText };
        json = Object.assign(json, jsonObject);
      }
      if (this.values.pcImage) {
        let pcImage = this.values.pcImage;
        this.log('Image: %s', pcImage);
        jsonObject = { image: pcImage };
        json = Object.assign(json, jsonObject);
      }
      if (this.values.sound !== 'default') {
        jsonObject = { sound: this.values.sound };
        json = Object.assign(json, jsonObject);
      }
      if (this.values.devices) {
        jsonObject = { devices: this.values.devices };
        json = Object.assign(json, jsonObject);
      }

      this.log('Url: %s, JSON: %j', this.pcClient.url, json);
      this.useReboot = false;

      const response = await this.pcClient
        .post(this.notification, json)
        .then((res) => {
          json = res.body;
          this.log('Notification sent: %s', json.message);
        })
        .catch((err) => {
          const str = String(err);
          this.debug(str);
          if (err.response) {
            this.warn(
              'Notification failed: %s - %s',
              err.response.statusCode,
              err.response.error
            );
          }
        });

      this.log('Starting the mute timer for %d s', this.mute);
      this.timeUp = false;
      try {
        await wait(this.mute * 1000, { signal: this.signal });
        this.log('Mute time is up!');
        this.timeUp = true;
        this.values.on = false;
        this.switchOn = false;
      } catch (err) {
        this.warn('The mute timer was aborted!');
      }
    }
  }
}

module.exports = SwitchService;


lib/PcPlatform.js
// homebridge-pushcutter/lib/PcPlatform.js
// Copyright © 2021-2022 Kenneth Jagenheim. All rights reserved.
//
// Homebridge plugin for Pushcut notifications.

'use strict';

const events = require('events');
const homebridgeLib = require('homebridge-lib');
const PcAccessory = require('./PcAccessory');
const PT = require('./PcTypes');
const PcTypes = PT.PcTypes;

class PcPlatform extends homebridgeLib.Platform {
  constructor(log, configJson, homebridge) {
    super(log, configJson, homebridge);
    //    this.once('heartbeat', this.init)
    this.config = {
      name: 'Pushcutter',
    };
    let UUIDGen = homebridge.hap.uuid;
    this.notificationIds = [];
    this.pushcutDevices = [];
    this.pc = new PcTypes(homebridge);
    let useApi = true;
    let host = 'api.pushcut.io';
    let secret = '/v1';
    let command = '/notifications';
    let headers = Object();

    const optionParser = new homebridgeLib.OptionParser(
      this.config,
      true
    );
    optionParser
      .stringKey('name')
      .stringKey('platform')
      .objectKey('api')
      .boolKey('selectApi')
      .stringKey('apiCode')
      .stringKey('locale')
      .arrayKey('notifications')
      .on('userInputError', (message) => {
        this.warn('config.json: %s', message);
      });
    try {
      optionParser.parse(configJson);
      if (this.config.notifications.length === 0) {
        this.warn('config.json: no notifications');
      }

      if (this.config.api.selectApi == 'secret') {
        useApi = false;
        secret = '/' + this.config.api.apiCode;
      } else {
        headers = { 'API-Key': this.config.api.apiCode };
      }
      this.log('API URL is "%s"', host + secret + command);
      this.pcClient = new homebridgeLib.HttpClient({
        https: true,
        host: host,
        headers: headers,
        json: true,
        maxSockets: 1,
        path: secret + command,
        timeout: 5, //this.config.timeout,
        validStatusCodes: [200, 401, 403, 404],
      });

      this.pcClient
        .on('error', (error) => {
          this.debug(
            'Pushcut request %d: Request = %s, Resource = %s',
            error.request.id,
            error.request.method,
            error.request.resource
          );
          this.warn(
            'Pushcut request %d: Error = %d, %s',
            error.request.id,
            error.statusCode,
            error.statusMessage
          );
        })
        .on('request', (request) => {
          this.debug(
            'Pushcut request %d: Request = %s, Resource = %s',
            request.id,
            request.method,
            request.resource
          );
          this.vdebug(
            'Pushcut request %d: Request = %s, URL = %s',
            request.id,
            request.method,
            request.url
          );
        })
        .on('response', (response) => {
          this.vdebug(
            'Pushcut request %d: Response = %j',
            response.request.id,
            response.body
          );
          this.debug(
            'Pushcut request %d: Response = %d %s',
            response.request.id,
            response.statusCode,
            response.statusMessage
          );
        });
    } catch (error) {
      this.error(error);
    }

    let emitter = new events.EventEmitter();
    // Devices and notifications can only be fetched when the API is used
    if (useApi) {
      // Get devices
      let restorePath = this.pcClient.path;
      this.pcClient.path = secret + '/devices';
      this.getDevices()
        .then((deviceArray) => {
          this.pushcutDevices = deviceArray;
          this.log('Devices: %s', deviceArray);
        })
        .catch((error) => {
          this.warn('Cannot get devices from Pushcut: %s', error);
        });

      // Check notifications
      this.pcClient.path = restorePath;
      this.getNotifications()
        .then((notificationIds) => {
          this.notificationIds = notificationIds;
          this.log('Notifications: %s', notificationIds);
          emitter.emit('notifications');
        })
        .catch((error) => {
          this.warn(
            'Cannot get notification names from Pushcut: %s',
            error
          );
        });
    } else {
      emitter.emit('notifications');
    }

    this.pushcutAccessories = {};
    const validNotifications = [];
    this.debug('Waiting for notifications');
    emitter.on('notifications', () => {
      this.debug('Got notifications');
      this.once('heartbeat', this.init);
      for (const i in this.config.notifications) {
        const pushcutSwitch = this.config.notifications[i];
        const config = {};
        const optionParser = new homebridgeLib.OptionParser(
          config,
          true
        );
        optionParser
          .stringKey('switchName')
          .stringKey('notificationName')
          .boolKey('useWebhook')
          .stringKey('webhook')
          .stringKey('pcTitle')
          .stringKey('pcText')
          .stringKey('pcImage')
          .intKey('mute', 1, 60)
          .boolKey('startOnReboot')
          .stringKey('rebootTitle')
          .stringKey('rebootText')
          .stringKey('sound')
          .stringKey('devices')
          .boolKey('useConfig')
          .on('userInputError', (error) => {
            this.warn('config.json: notifications[%d]: %s', i, error);
          });
        optionParser.parse(pushcutSwitch);
        config.uuid = UUIDGen.generate(config.switchName);
        this.log('Found switch: %s', config.switchName);
        if (config.useWebhook) {
          try {
            let fullUrl = new URL(config.webhook);
            let hookUrl = fullUrl.href;
            let hookHost = fullUrl.host;
            let hookPathname = fullUrl.pathname;
            // Remove trailing slash
            let trimmedPath =
              hookPathname.charAt(hookPathname.length - 1) == '/'
                ? hookPathname.slice(0, -1)
                : hookPathname;

            let urlPaths = trimmedPath.split('/');
            config.notificationName = urlPaths[urlPaths.length - 1];
            if (config.webhook != hookUrl) {
              this.log(
                'Webhook URL automatically adjusted from "%s" to "%s"',
                config.webhook,
                hookUrl
              );
              config.webhook = hookUrl;
            }
            this.log(
              'Using Webhook URL "%s" for %s',
              hookUrl,
              config.switchName
            );
          } catch (error) {
            this.warn(
              'Error parsing webhook %s: %s',
              config.webhook,
              error.message
            );
            this.warn('Skipping switch %s...', config.switchName);
            continue;
          }
        }

        // Check if the notification is available from Pushcut
        if (this.notificationIds) {
          let included = this.notificationIds.includes(
            config.notificationName
          );
          if (included) {
            this.log(
              'Confirmed that Pushcut includes notification %s',
              config.notificationName
            );
          } else {
            if (config.useWebhook) {
              this.warn(
                'Webhook notification name not found in Pushcut: %s',
                config.notificationName
              );
            } else {
              this.warn(
                'Notification name not found in Pushcut: %s',
                config.notificationName
              );
            }
          }
        } else {
          this.log(
            'Notification name cannot be validated with Pushcut, must use API key'
          );
        }

        // Check that devices are available
        if (config.devices) {
          let devices = config.devices.split(',');
          let devicesAvailable = true;
          devices.forEach((device) => {
            let included = this.pushcutDevices.includes(device);
            if (!included) {
              devicesAvailable = false;
            }
          });
          if (!devicesAvailable) {
            this.warn(
              'Devices not available in Pushcut, will be ignored'
            );
            config.devices = '';
          }
        }

        // Check image URL
        if (config.pcImage) {
          try {
            let fullImageUrl = new URL(config.pcImage);
            let imageUrl = fullImageUrl.href;
            this.log('Found valid image URL: %s', imageUrl);
          } catch (error) {
            this.warn(
              'Error parsing image URL %s: %s',
              imageUrl,
              error.message
            );
            this.log('Image may not be shown for', config.switchName);
          }
        }
        config.url = this.config.api.url + config.notificationName;
        config.pcClient = this.pcClient;
        //}
        validNotifications.push(config);
      }
      this.config.notifications = validNotifications;
      this.debug('config: %j', this.config);
    });
  }

  async init(beat) {
    this.debug('Init started');
    const jobs = [];
    for (const pushcutSwitch of this.config.notifications) {
      const switchParams = {
        name: pushcutSwitch.switchName,
        id: pushcutSwitch.uuid,
        manufacturer: 'Homebridge',
        model: 'Pushcut-' + pushcutSwitch.switchName,
        category: this.Accessory.Categories.Switch,
        api: this.config.api,
        //        url: pushcutSwitch.url,
        notification: '/' + pushcutSwitch.notificationName,
        pcClient: pushcutSwitch.pcClient,
        useWebhook: pushcutSwitch.useWebhook,
        webhook: pushcutSwitch.webhook,
        pcTitle: pushcutSwitch.pcTitle || '',
        pcText: pushcutSwitch.pcText || '',
        pcImage: pushcutSwitch.pcImage || '',
        mute: pushcutSwitch.mute,
        startOnReboot: pushcutSwitch.startOnReboot,
        rebootTitle: pushcutSwitch.rebootTitle || '',
        rebootText: pushcutSwitch.rebootText || '',
        sound: pushcutSwitch.sound || 'default',
        devices: pushcutSwitch.devices || '',
        useConfig: pushcutSwitch.useConfig,
        pushcutDevices: this.pushcutDevices,
      };
      const switchAccessory = new PcAccessory(this, switchParams);
      jobs.push(events.once(switchAccessory, 'initialised'));
      this.pushcutAccessories[pushcutSwitch] = switchAccessory;
    }
    for (const job of jobs) {
      await job;
    }
    this.debug('Platform initialised');
    this.emit('initialised');
  }

  async getNotifications() {
    try {
      let notificationArray = [];
      const notifications = await this.pcClient.get();
      const json = await notifications.body;
      json.forEach((element) => {
        notificationArray.push(element.id);
      });
      return notificationArray;
    } catch (err) {
      this.warn(
        'Cannot check notifications:',
        err.statusCode,
        err.body.error
      );
    }
  }

  async getDevices() {
    try {
      let deviceArray = [];
      const devices = await this.pcClient.get();
      const json = await devices.body;
      json.forEach((element) => {
        deviceArray.push(element.id);
      });
      return deviceArray;
    } catch (err) {
      this.warn('Cannot check devices:', err.statusCode);
    }
  }
}

module.exports = PcPlatform;
